controllers/usersignup/usersignup_controller.go

```diff
@@ -107,11 +120,24 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
 	}
 	logger = logger.WithValues("username", userSignup.Spec.Username)
 
+	// Don't reconcile while migration is in progress
+	// TODO delete this after migration is complete
+	_, ok := userSignup.Annotations[migrationInProgressAnnotationName]
```

Comment:
It shouldn't be blocked. 
There are two phases:
1. creation of a new UserSignup
2. deletion of the old one

the deletion can happen only when the `migrationInProgressAnnotationName` is not present.
The first phase consists of two steps
1. actual creation of the new UserSignup
2. update of the `UserSignup.status` and removal of the annotation `migrationInProgressAnnotationName`

if the second step fails (for example because of a conflict) then the controller returns an error and got reconciled again. in the next run it updates the status & removes the annotation. Since the UserSignup predicates checks also for the changes in annotation, then the controller should be triggered for the "new" UserSignup which doesn't have the annotation anymore.

This annotation and the check was introduced so the controller is not triggered before the "new" UserSignup has all needed status conditions set.

Correct me if I missed some detail

Reply:
but the error is returned during the reconciliation of the "old" UserSignup:
* "Old UserSignup" is reconciled
   1. it creates "new UserSignup" with the annotation
   2. it tries to update the status
   3. it fails and returns an error
* "New UserSignup" is reconciled based on its creation
    1. it exits because the annotation is still present
* "Old UserSignup" is reconciled because the previous was returned with an error
   1. it doesn't create anything because the new is already present
   2. it tries to update the status of the "new UserSignup"
   3. it removes the annotation
* "New UserSignup" is reconciled based on the removal of the annotation
   1. deletes the "old UserSignup"

Reply:
Ah, you're right. I was in the habit of thinking of reconciliations on the same resource but in this case we're migrating to a new resource altogether to please disregard what I said. Sorry about that.